File: frontend/ideacapital/lib/models/invention.dart
import 'package:json_annotation/json_annotation.dart';

part 'invention.g.dart';

/// Mirrors the canonical InventionSchema.json.
/// This is the Dart representation of the shared schema.
@JsonSerializable(explicitToJson: true)
class Invention {
  @JsonKey(name: 'invention_id')
  final String inventionId;
  final String status;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime? updatedAt;
  @JsonKey(name: 'creator_id')
  final String creatorId;
  @JsonKey(name: 'social_metadata')
  final SocialMetadata socialMetadata;
  @JsonKey(name: 'technical_brief')
  final TechnicalBrief? technicalBrief;
  @JsonKey(name: 'risk_assessment')
  final RiskAssessment? riskAssessment;
  final Funding? funding;
  @JsonKey(name: 'blockchain_ref')
  final BlockchainRef? blockchainRef;

  Invention({
    required this.inventionId,
    required this.status,
    required this.createdAt,
    this.updatedAt,
    required this.creatorId,
    required this.socialMetadata,
    this.technicalBrief,
    this.riskAssessment,
    this.funding,
    this.blockchainRef,
  });

  factory Invention.fromJson(Map<String, dynamic> json) =>
      _$InventionFromJson(json);
  Map<String, dynamic> toJson() => _$InventionToJson(this);
}

@JsonSerializable(explicitToJson: true)
class SocialMetadata {
  @JsonKey(name: 'display_title')
  final String displayTitle;
  @JsonKey(name: 'short_pitch')
  final String shortPitch;
  @JsonKey(name: 'virality_tags')
  final List<String>? viralityTags;
  @JsonKey(name: 'media_assets')
  final MediaAssets? mediaAssets;

  SocialMetadata({
    required this.displayTitle,
    required this.shortPitch,
    this.viralityTags,
    this.mediaAssets,
  });

  factory SocialMetadata.fromJson(Map<String, dynamic> json) =>
      _$SocialMetadataFromJson(json);
  Map<String, dynamic> toJson() => _$SocialMetadataToJson(this);
}

@JsonSerializable()
class MediaAssets {
  @JsonKey(name: 'hero_image_url')
  final String? heroImageUrl;
  @JsonKey(name: 'explainer_video_url')
  final String? explainerVideoUrl;
  @JsonKey(name: 'thumbnail_url')
  final String? thumbnailUrl;
  final List<String>? gallery;

  MediaAssets({
    this.heroImageUrl,
    this.explainerVideoUrl,
    this.thumbnailUrl,
    this.gallery,
  });

  factory MediaAssets.fromJson(Map<String, dynamic> json) =>
      _$MediaAssetsFromJson(json);
  Map<String, dynamic> toJson() => _$MediaAssetsToJson(this);
}

@JsonSerializable(explicitToJson: true)
class TechnicalBrief {
  @JsonKey(name: 'technical_field')
  final String? technicalField;
  @JsonKey(name: 'background_problem')
  final String? backgroundProblem;
  @JsonKey(name: 'solution_summary')
  final String? solutionSummary;
  @JsonKey(name: 'core_mechanics')
  final List<CoreMechanic>? coreMechanics;
  @JsonKey(name: 'novelty_claims')
  final List<String>? noveltyClaims;
  @JsonKey(name: 'hardware_requirements')
  final List<String>? hardwareRequirements;
  @JsonKey(name: 'software_logic')
  final String? softwareLogic;

  TechnicalBrief({
    this.technicalField,
    this.backgroundProblem,
    this.solutionSummary,
    this.coreMechanics,
    this.noveltyClaims,
    this.hardwareRequirements,
    this.softwareLogic,
  });

  factory TechnicalBrief.fromJson(Map<String, dynamic> json) =>
      _$TechnicalBriefFromJson(json);
  Map<String, dynamic> toJson() => _$TechnicalBriefToJson(this);
}

@JsonSerializable()
class CoreMechanic {
  final int step;
  final String description;

  CoreMechanic({required this.step, required this.description});

  factory CoreMechanic.fromJson(Map<String, dynamic> json) =>
      _$CoreMechanicFromJson(json);
  Map<String, dynamic> toJson() => _$CoreMechanicToJson(this);
}

@JsonSerializable(explicitToJson: true)
class RiskAssessment {
  @JsonKey(name: 'potential_prior_art')
  final List<PriorArt>? potentialPriorArt;
  @JsonKey(name: 'feasibility_score')
  final int? feasibilityScore;
  @JsonKey(name: 'missing_info')
  final List<String>? missingInfo;

  RiskAssessment({
    this.potentialPriorArt,
    this.feasibilityScore,
    this.missingInfo,
  });

  factory RiskAssessment.fromJson(Map<String, dynamic> json) =>
      _$RiskAssessmentFromJson(json);
  Map<String, dynamic> toJson() => _$RiskAssessmentToJson(this);
}

@JsonSerializable()
class PriorArt {
  final String? source;
  @JsonKey(name: 'patent_id')
  final String? patentId;
  @JsonKey(name: 'similarity_score')
  final double? similarityScore;
  final String? notes;

  PriorArt({this.source, this.patentId, this.similarityScore, this.notes});

  factory PriorArt.fromJson(Map<String, dynamic> json) =>
      _$PriorArtFromJson(json);
  Map<String, dynamic> toJson() => _$PriorArtToJson(this);
}

@JsonSerializable()
class Funding {
  @JsonKey(name: 'goal_usdc')
  final double? goalUsdc;
  @JsonKey(name: 'raised_usdc')
  final double? raisedUsdc;
  @JsonKey(name: 'backer_count')
  final int? backerCount;
  @JsonKey(name: 'min_investment_usdc')
  final double? minInvestmentUsdc;
  @JsonKey(name: 'royalty_percentage')
  final double? royaltyPercentage;
  final DateTime? deadline;
  @JsonKey(name: 'token_supply')
  final int? tokenSupply;

  Funding({
    this.goalUsdc,
    this.raisedUsdc,
    this.backerCount,
    this.minInvestmentUsdc,
    this.royaltyPercentage,
    this.deadline,
    this.tokenSupply,
  });

  factory Funding.fromJson(Map<String, dynamic> json) =>
      _$FundingFromJson(json);
  Map<String, dynamic> toJson() => _$FundingToJson(this);
}

@JsonSerializable()
class BlockchainRef {
  @JsonKey(name: 'nft_contract_address')
  final String? nftContractAddress;
  @JsonKey(name: 'nft_token_id')
  final String? nftTokenId;
  @JsonKey(name: 'royalty_token_address')
  final String? royaltyTokenAddress;
  @JsonKey(name: 'crowdsale_address')
  final String? crowdsaleAddress;
  @JsonKey(name: 'chain_id')
  final int? chainId;
  @JsonKey(name: 'ipfs_metadata_cid')
  final String? ipfsMetadataCid;

  BlockchainRef({
    this.nftContractAddress,
    this.nftTokenId,
    this.royaltyTokenAddress,
    this.crowdsaleAddress,
    this.chainId,
    this.ipfsMetadataCid,
  });

  factory BlockchainRef.fromJson(Map<String, dynamic> json) =>
      _$BlockchainRefFromJson(json);
  Map<String, dynamic> toJson() => _$BlockchainRefToJson(this);
}

File: brain/src/models/invention.py
"""
Invention Pydantic models - Python representation of InventionSchema.json.

These models are the Python-side contract for the shared schema.
They mirror the canonical JSON Schema in /schemas/InventionSchema.json.
"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


class MediaAssets(BaseModel):
    hero_image_url: Optional[str] = None
    explainer_video_url: Optional[str] = None
    thumbnail_url: Optional[str] = None
    gallery: list[str] = Field(default_factory=list)


class SocialMetadata(BaseModel):
    display_title: str = Field(max_length=60)
    short_pitch: str = Field(max_length=280)
    virality_tags: list[str] = Field(default_factory=list)
    media_assets: Optional[MediaAssets] = None


class CoreMechanic(BaseModel):
    step: int
    description: str


class TechnicalBrief(BaseModel):
    technical_field: Optional[str] = None
    background_problem: Optional[str] = None
    solution_summary: Optional[str] = None
    core_mechanics: list[CoreMechanic] = Field(default_factory=list)
    novelty_claims: list[str] = Field(default_factory=list)
    hardware_requirements: list[str] = Field(default_factory=list)
    software_logic: Optional[str] = None


class PriorArt(BaseModel):
    source: str
    patent_id: str
    similarity_score: float = Field(ge=0, le=1)
    notes: str


class RiskAssessment(BaseModel):
    potential_prior_art: list[PriorArt] = Field(default_factory=list)
    feasibility_score: int = Field(ge=1, le=10, default=5)
    missing_info: list[str] = Field(default_factory=list)


class Funding(BaseModel):
    goal_usdc: Optional[float] = None
    raised_usdc: float = 0
    backer_count: int = 0
    min_investment_usdc: Optional[float] = None
    royalty_percentage: Optional[float] = None
    deadline: Optional[datetime] = None
    token_supply: Optional[int] = None


class BlockchainRef(BaseModel):
    nft_contract_address: Optional[str] = None
    nft_token_id: Optional[str] = None
    royalty_token_address: Optional[str] = None
    crowdsale_address: Optional[str] = None
    chain_id: Optional[int] = None
    ipfs_metadata_cid: Optional[str] = None


class InventionDraft(BaseModel):
    """Full invention model matching the canonical InventionSchema.json."""

    invention_id: str
    status: str = "DRAFT"
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None
    creator_id: str
    social_metadata: SocialMetadata
    technical_brief: Optional[TechnicalBrief] = None
    risk_assessment: Optional[RiskAssessment] = None
    funding: Optional[Funding] = None
    blockchain_ref: Optional[BlockchainRef] = None
